### PR TITLE
models: Add FieldIterator type

### PR DESCRIPTION
**Pull request by joelegasse**
The `FieldIterator` is used to scan over the fields of a point, providing information, and delaying parsing/decoding the value until it is needed.

This change uses this new type to avoid the allocation of a map for the fields which is then thrown away as soon as the points get converted into columns within the datastore.

There are also some smaller changes that would accumulate allocations over time, mostly around strings.

This should definitely be tested for correctness before being merged... The tests pass, but they might not cover all edge cases.

###### Required for all non-trivial PRs
- [x] Rebased/mergable
- [x] Tests pass
- [ ] CHANGELOG.md updated

### CODE REVIEW

File: models/points.go [hunk 1]
```diff
@@ -89,6 +89,33 @@ type Point interface {
...
+	Integer FieldType = iota
+	Float
+	Boolean
+	String
+	Empty
+)
+
+type FieldIterator interface {
+	Next() bool
+	FieldKey() []byte
```
Comment (jwilder) [inline models/points.go]: `Point` implement this interface and already has `Key` defined which would conflict with intent `FieldKey`

File: models/points.go [hunk 2]
```diff
@@ -89,6 +89,33 @@ type Point interface {
...
+	Type() FieldType
+	StringValue() string
```
Comment (jwilder) [inline models/points.go]: `Point` also already has a `String` func defined which would conflict with this.
